pkg/fqdn/service/service.go

```diff
@@ -100,31 +141,118 @@ type PolicyUpdater interface {
 	UpdatePolicyRules(map[identity.NumericIdentity]policy.SelectorPolicy, bool) error
 }
 
+var closedWatchChannel = func() <-chan struct{} {
+	ch := make(chan struct{})
+	close(ch)
+	return ch
+}()
+
 // StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
 // SDP calls this method to subscribe to DNS policies
-// For each stream, we start a goroutine to receive the DNS policies ACKs
+// For each stream, we subscribe to the changes in the policy rules table and send the current state of the DNS rules to the client.
 // The flow of the method is as follows:
-// 1. Add the stream to the map( called by the client i.e SDP)
-// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
-// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
-// 4. Wait for the context to be done
-// Note: this method is left empty on purpose and will be update with the actual implementation in the future PRs for the standalone DNS proxy
+//  1. Send the current state of the DNS rules to the client.
+//  2. Subscribe to the changes in the policy rules table and wait for changes.
+//  3. For each change in the policy rules table, send the current state of the DNS rules to the client.
+//  4. If the stream context is done, return.
 func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
-	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
-	stream.Send(&pb.PolicyState{RequestId: "test"})
+	streamCtx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	limiter := rate.NewLimiter(time.Second, 1)
+	defer limiter.Stop()
+
+	rulesWatch := closedWatchChannel
+
+	for {
+
+		select {
+		case <-streamCtx.Done():
+			return streamCtx.Err()
+		case <-rulesWatch:
+			// If there are changes in the policy rules table, we will send the current state
+			// of the DNS policies to the client.
+			rules, watch := s.policyRulesTable.AllWatch(s.db.ReadTxn())
+			if err := s.sendAndRecvAckForDNSPolicies(stream, rules); err != nil {
+				return err
+			}
+			rulesWatch = watch
+		}
+		// Limit the rate at which we send the full snapshots
+		if err := limiter.Wait(streamCtx); err != nil {
+			return err
+		}
+	}
+
+}
+
+// sendAndRecvAckForDNSPolicies sends the current state of the DNS policies to the client
+// and waits for the ACK from the client.
+func (s *FQDNDataServer) sendAndRecvAckForDNSPolicies(stream pb.FQDNData_StreamPolicyStateServer, rules iter.Seq2[policyRules, statedb.Revision]) error {
+	requestID := uuid.New().String()
+	policyState := &pb.PolicyState{
+		RequestId: requestID,
```

Comment:
Don't you need to set the rules here :-) ?

Reply:
Yes absolutely. I didn't update it intentionally because it is supposed to be added once I create the next PR for `updatePolicyRules`.
I think it will require more discussion so kept it separate. I added a comment for now on this function.